agenthub/codeact_agent/codeact_agent.py

```diff
@@ -194,27 +215,148 @@ def step(self, state: State) -> Action:
+            print('No of tokens, ' + str(self.llm.get_token_count(messages)) + '\n')
+            try:
+                if self.llm.is_over_token_limit(messages):
+                    raise TokenLimitExceededError()
+                response = self.llm.completion(
+                    messages=messages,
+                    stop=[
+                        '</execute_ipython>',
+                        '</execute_bash>',
+                        '</execute_browse>',
+                    ],
+                    temperature=0.0,
+                )
+            except (ContextWindowExceededError, TokenLimitExceededError):
+                # Handle the specific exception
+                print('An error occurred: ')
+                attempt += 1
+                # If we got a context alert, try trimming the messages length, then try again
+                if self.llm.is_over_token_limit(messages):
+                    # A separate call to run a summarizer
+                    self.condense(state=state)
+                    # Try step again
+                else:
+                    print('step() failed with an unrecognized exception:')
+                    raise ContextWindowLimitExceededError()
+
+            # TODO: Manage the response for exception.
         return self.action_parser.parse(response)
 
+    def condense(
+        self,
+        state: State,
+    ):
+        # Start past the system message, and example messages.,
+        # and collect messages for summarization until we reach the desired truncation token fraction (eg 50%)
+        # Do not allow truncation  for in-context examples of function calling
+        history: ShortTermHistory = state.history
+        messages = self._get_messages(state=state)
+        token_counts = [self.llm.get_token_count([message]) for message in messages]
+        message_buffer_token_count = sum(
+            token_counts[2:]
+        )  # no system and example message
+        MESSAGE_SUMMARY_TRUNC_TOKEN_FRAC = 0.75
+        desired_token_count_to_summarize = int(
+            message_buffer_token_count * MESSAGE_SUMMARY_TRUNC_TOKEN_FRAC
+        )
+
+        candidate_messages_to_summarize = []
+        last_summarized_event_id = history.last_summarized_event_id
+        tokens_so_far = 0
+        for event in history.get_events():
+            if isinstance(event, AgentSummarizeAction):
+                action_message = get_action_message(event)
+                if action_message:
+                    candidate_messages_to_summarize.append(action_message)
+                    tokens_so_far += self.llm.get_token_count([action_message])
+                observation_message = get_observation_message(event)
+                if observation_message:
+                    candidate_messages_to_summarize.append(observation_message)
+                    tokens_so_far += self.llm.get_token_count([observation_message])
+                continue
+            else:
```

Comment:
I'm imagine a `AgentSummarizeAction` will act like a blocker - it will allow us to ignore all the previous messages that are NOT `condensable` (e.g., we will assume that they are already summarized in the action). So we probably shouldn't include `condensable` actions/observations BEFORE the *last* summarize action. WDYT?

Reply:
Corrected it